### ISSUE TITLE
Data corruption when data (shapefiles) is stored on a Windows Network share and OS is windows 10 enterprise ltsc 2019 / SMB 2

### ISSUE THREAD

**Issue by thbaumann** [qgis/QGIS#45372]
### What is the bug or the crash?

Sometimes feature geometry turns into nullgeometries when saving layer edits. In the attached screencast you can see this.

https://user-images.githubusercontent.com/20082727/135811059-74df7e42-f2a8-45ee-98d0-a8f555d1931e.mp4

It seems to only affect shapefiles stored on network shares. There  I can reproduce this error as often as I want. With shapefiles stored on local drives I could not reproduce the error yet. The QGIS client I used to test is on a Windows10 machine.

( The but also occurs if you start QGIS with the --noplugins parameter. )

### Steps to reproduce the issue
1. Use a PC with Win10 LSB2019
2. Copy the three attached test-shapefiles to a network share
3. load the shapefiles into QGIS
4. ( I turned on snapping but this step is probably not neccassary ) 
5. Turn the testdata-layer into editing mode
6. use the offset-tool to move one line segment as shown in the screencast (to the position of the target shapefile)
this seems to trigger the error of the cut feature tool
7. cut the testdata-shapefile along the line of the shapefile cutline.shp and save the changes
8. cut the testdata-shapefile a second time along the line of the shapefile cutline.shp and save the changes
9. now the cut off segments should turn into nullegeoms as shown in the screencast
10. from now on every cut off segment will turn into a nullgeom
11. if you reload the layer or do a iface.activeLayer().dataProvider().forceReload() once the bug is triggered you can cut off line segments as expected again. If you pan/zoom in or out (causing a reload of the data) during the steps to reproduce the bug the error will not occur.

### Versions

3.16.11, 3.18.3, 3.20.3, 3.22.6, 3.24.2

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by thbaumann** [comment]
https://user-images.githubusercontent.com/20082727/135810870-bea396c7-d569-45a3-8f83-3c6b887a5d3f.mp4
[testdata.zip](https://github.com/qgis/QGIS/files/7276235/testdata.zip)

**Comment by uclaros** [comment]
Can you check that you have write permissions on the samba share?

**Comment by thbaumann** [comment]
@uclaros : Yes I have write permissions. everytime I press save the files are saved on the network share but instead of geomeries there are nullgeoms for the cut off segments.

**Comment by gioman** [comment]
> Yes I have write permissions.

@thbaumann so that does never happen if the layers are on a local disk?

**Comment by thbaumann** [comment]
@gioman : i have done a lot of tests to narrow down this bug and so far i
haven't been able to reproduce the error on a local disk. On windows shares
i can reproduce it every time I try.

Giovanni Manghi ***@***.***> schrieb am Mo., 4. Okt. 2021,
20:42:

**Comment by gioman** [comment]
> @gioman : i have done a lot of tests to narrow down this bug and so far i haven't been able to reproduce the error on a local disk. On windows shares i can reproduce it every time I try.

@thbaumann does it happen with GPKGs or any other format other than shp?

**Comment by thbaumann** [comment]
No the geopackage acts different. In the shapefile an empty geometry will turn into a nullgeometry if you save the shapefile.
The occurence of a nullgeometry in the layer then seems to trigger the cut feature tool bug.

The geopackage can handle with empty linestrings like
`MultiLineString ((431134.10965812 5310060.77920309, ........ , 431345.46394564 5310058.64943521),LineString EMPTY)` and there the bug does not occur.

**Comment by thbaumann** [comment]
Perhaps this issue has also to do with the circumstance that geometries turn sometimes into nullgeometries when shapefiles are stored on windows samba shares: https://github.com/qgis/QGIS/issues/41964

**Comment by thbaumann** [comment]
Any thoughts in which cpp-File / part of the code i should start to search the reasons for this bug ? QgsVectorLayerEditBuffer::commitChanges ? 
Until I save the layer edits the cut off line segments still exists. Pressing the mActionSaveLayerEdits Button is the "trigger" when the geometry turns into nulllgeoms / empty geometries.

**Comment by thbaumann** [comment]
As a workaround you can use afterCommitChanges signal and do a forceReload every time geometries turn into nullgeoms. This works so far as a workaround for our testcase but is obviously not fixing the problem itself.
```
try:
    layer.beforeCommitChanges.disconnect()
    layer.afterCommitChanges.disconnect()

except:
    pass    
layer = iface.activeLayer()

def little_helper(zeitpunkt):
        it = layer.getFeatures()
        nullgeoms=[]
        for feature in it:
            geom = feature.geometry()
            if feature.geometry().isNull():
                nullgeoms.append(feature.id())
                
        if len(nullgeoms)>0:
            layer.dataProvider().forceReload()
            print(zeitpunkt+": force reload")
            print(nullgeoms)
            try:
                layer.selectByIds([s for s in nullgeoms])
                iface.actionZoomToSelected().trigger();
            except:
                pass

layer.beforeCommitChanges.connect(lambda: little_helper('before'))

layer.afterCommitChanges.connect(lambda: little_helper('after'))
```

**Comment by thbaumann** [comment]
For my testcase the bug is not reproducible anymore if I set following reistry keys to 0 : 
DirectoryCacheLifetime, FileInfoCacheLifetime and FileNotFoundCacheLifetime ( HKEY_LOCAL_MACHINE\System\CurrentControlSet\Services\Lanmanworkstation\Parameters ) .
ESRI also had problems with the SMB redirector cache in the past and the Win10LTSSC2019 seems to have a similar problem which can be fixed by setting these registry keys.
https://support.esri.com/en/technical-article/000011778

**Comment by thbaumann** [comment]
Bug still exists in QGIS 3.28 /3.30 /  QGIS 3.32

**Comment by thbaumann** [comment]
Bug still exists in QGIS 3.28 /3.30 / QGIS 3.32
